doc/remediation-templating.md

```diff
@@ -0,0 +1,207 @@
+     fixed value
+   * During a scan where an administrator set custom variables in the tailored profile, it
+     is possible to create `ComplianceRemediation` objects based on the given custom values
+   * As a content writer, it is possible to use XCCDF value from rules in Kubernetes remediation
+
+## High-level overview
+
+The compliance operator allows an administrator to extend a pre-existing profile, 
+to enable, disable rules and set values that come from the ProfileBundle. When 
+this happens, a CR `TailoredProfile` will be generated. Currently the `ComplianceScan`
+will use the tailored rules and variables from `TailoredProfile` CR, however,
+the `ComplianceRemediation` will now take custom values from `TailoredProfile`.
+Here we are going to implement that feature.
+
+### ComplianceAsCode remediation rule format
+
+
+In ComplianceAsCode, when the SCAP remediation content is generated, normally we will 
+have fixed remediation. In this proposal, we are going to change the remediation rule
+formate for the Kubernetes to include a marked reference variable to reference related
+XCCDF values from rules. 
+
+The following example is parts of two YAML files from ComplianceAsCode content, 
+the first one is the rule, and the second is the remediation:
+```yaml
+documentation_complete: true
+title: 'Set SSH Idle Timeout Interval'
+...
+requires:
+    - sshd_set_keepalive
+
+ocil_clause: 'it is commented out or not configured properly'
+
+ocil: |-
+    Run the following command to see what the timeout interval is:
+    <pre>$ sudo grep ClientAliveInterval /etc/ssh/sshd_config</pre>
+    If properly configured, the output should be:
+    <pre>ClientAliveInterval {{{ xccdf_value("sshd_idle_timeout_value") }}}</pre>
+...
+```
+As showing in the rule YAML file above, an XCCDF value for SSH Idle Timeout Interval is set.
+
+In the remediation YAML file, we can modify that Marco to pass the name of the 
+XCCDF value `VAR_SSHD_IDLE_TIMEOUT_VALUE`, and then modify the related `ClientAliveCountMax`
+values to `VAR_SSHD_IDLE_TIMEOUT_VALUE`:
+```yaml
+---
+# platform = multi_platform_ocp,multi_platform_rhcos
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
+{{{ kubernetes_sshd_set("VAR_SSHD_IDLE_TIMEOUT_VALUE") }}}
+```
+
+And the machine config object data will be something like below before it gets url-encoded.
+```jinja
+...
+PrintLastLog yes
+#TCPKeepAlive yes
+PermitUserEnvironment no
+Compression no
+ClientAliveInterval 600
+ClientAliveCountMax VAR_SSHD_IDLE_TIMEOUT_VALUE
+
+#UseDNS no
+#PidFile /var/run/sshd.pid
+#MaxStartups 10:30:100
+#PermitTunnel no
+...
+```
+Since the url encode doesn't encode ASCII alpha-numerical nor underscore characters, and
+we can later change the place holder after the url encoding from `VAR_SSHD_IDLE_TIMEOUT_VALUE`
+to `{{sshd_idle_timeout_value}}`, so that it can be post-process them with the appropriate
```

Comment:
Can you please explicitly mention which component will perform this substitution?

Reply:
I am thinking about add a function in the Macro to do that substitution, before we create the remediation xml file.

Reply:
Right, so from `VAR_SSHD_IDLE_TIMEOUT_VALUE` to `{{sshd_idle_timeout_value}}` it would be part of the content building. But it would be good to know where the substitution from `{{sshd_idle_timeout_value}}` to the actual value happens as well

Reply:
Yes, so after we have the remediation xml data with `{{some-value-need-get-replaced}}` , in the aggregator, before we create the remediation, we need to look up  `some-value-need-get-replaced` against the result `config-map` and we can then replace the value with the `Set-Value` in result `config-map`